core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1502,6 +1502,53 @@ private[spark] class DAGScheduler(
             failedStages += failedStage
             failedStages += mapStage
             if (noResubmitEnqueued) {
+              // If the map stage is INDETERMINATE, which means the map tasks may return
+              // different result when re-try, we need to re-try all the tasks of the failed
+              // stage and its succeeding stages, because the input data will be changed after the
+              // map tasks are re-tried.
+              // Note that, if map stage is UNORDERED, we are fine. The shuffle partitioner is
+              // guaranteed to be idempotent, so the input data of the reducers will not change even
+              // if the map tasks are re-tried.
+              if (mapStage.rdd.computingRandomLevel == RDD.RandomLevel.INDETERMINATE) {
+                def rollBackStage(stage: Stage): Unit = stage match {
+                  case mapStage: ShuffleMapStage =>
+                    val numMissingPartitions = mapStage.findMissingPartitions().length
+                    if (numMissingPartitions < mapStage.numTasks) {
+                      markStageAsFinished(
+                        mapStage,
+                        Some("preceding shuffle map stage with random output gets retried."),
+                        willRetry = true)
+                      mapOutputTracker.unregisterAllMapOutput(mapStage.shuffleDep.shuffleId)
+                      failedStages += mapStage
+                    }
+
+                  case resultStage =>
+                    val numMissingPartitions = resultStage.findMissingPartitions().length
+                    if (numMissingPartitions < resultStage.numTasks) {
```

Comment:
IIRC this can be a valid case - for example if result stage is being run on only a subset of partitions (first() for example) : so number of missing partitions can be legitimately > numTasks and still have missing relevant partitions.
I might be a bit rusty here though, +CC @markhamstra

Reply:
when run `first()`, I think `numTasks` will be 1, but `numPartitions` can be much larger.